Sources/Bond/AppKit/NSOutlineView+ObservableTree.swift

```diff
@@ -0,0 +1,152 @@
+                outlineView.removeItems(at: IndexSet(integer: at.item), inParent: parent, withAnimation: deleteAnimation ?? [])
+            case .update(let at):
+                let parent = rootNode[at.dropLast()]
+                let item = outlineView.child(at.item, ofItem: parent)
+                outlineView.reloadItem(item)
+            case .move(let from, let to):
+                let fromParent = rootNode[from.dropLast()]
+                let toParent = rootNode[to.dropLast()]
+                outlineView.moveItem(at: from.item, inParent: fromParent, to: to.item, inParent: toParent)
+            }
+        }
+
+        outlineView.endUpdates()
+    }
+}
+
+
+public extension SignalProtocol where Element: ObservableCollectionEventProtocol, Element.UnderlyingCollection: TreeNodeProtocol, Error == NoError {
+
+    @discardableResult
+    public func bind(to outlineView: NSOutlineView, animated: Bool = true, createCell: @escaping (Element.UnderlyingCollection, Element.UnderlyingCollection.Value?, NSTableColumn?, NSOutlineView) -> NSView?) -> Disposable {
+        let binder = OutlineViewBinder(measureCell: nil, createCell: createCell)
+        if !animated {
+            binder.deleteAnimation = nil
+            binder.insertAnimation = nil
+        }
+        return self.bind(to: outlineView, using: binder)
+    }
+
+    @discardableResult
+    public func bind(to outlineView: NSOutlineView, using binder: OutlineViewBinder<Element.UnderlyingCollection>) -> Disposable {
+        let dataSource = Property<Element.UnderlyingCollection?>(nil)
+        let disposable = CompositeDisposable()
+
+        disposable += outlineView.reactive.delegate.feed(
+            property: dataSource,
+            to: #selector(NSOutlineViewDelegate.outlineView(_:heightOfRowByItem:)),
+            map: { (dataSource: Element.UnderlyingCollection?, outlineView: NSOutlineView, item: Element.UnderlyingCollection.Value?) -> CGFloat in
+                guard let dataSource = dataSource else { return outlineView.rowHeight }
+                return binder.height(for: item, outlineView: outlineView, dataSource: dataSource) ?? outlineView.rowHeight
+            }
+        )
+
+        disposable += outlineView.reactive.delegate.feed(
+            property: dataSource,
+            to: #selector(NSOutlineViewDelegate.outlineView(_:viewFor:item:)),
+            map: { (dataSource: Element.UnderlyingCollection?, outlineView: NSOutlineView, tableColumn: NSTableColumn, item: Element.UnderlyingCollection.Value?) -> NSView? in
+                guard let dataSource = dataSource else { return nil }
+                return binder.cell(for: item, tableColumn: tableColumn, outlineView: outlineView, dataSource: dataSource)
+            }
+        )
+
+        disposable += outlineView.reactive.dataSource.feed(
+            property: dataSource,
+            to: #selector(NSOutlineViewDataSource.outlineView(_:numberOfChildrenOfItem:)),
```

Comment:
I'm seeing a crash in BNDInvocation that originates from this method:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = Fatal error: Unexpectedly found nil while unwrapping an Optional value
    frame #0: 0x00000001074f1fc0 libswiftCore.dylib`_swift_runtime_on_report
    frame #1: 0x0000000107535b31 libswiftCore.dylib`_swift_stdlib_reportFatalError + 113
    frame #2: 0x0000000107266856 libswiftCore.dylib`function signature specialization <Arg[1] = [Closure Propagated : reabstraction thunk helper from @callee_guaranteed (@unowned Swift.UnsafeBufferPointer<Swift.UInt8>) -> () to @callee_guaranteed (@unowned Swift.UnsafeBufferPointer<Swift.UInt8>) -> (@out ()), Argument Types : [@callee_guaranteed (@unowned Swift.UnsafeBufferPointer<Swift.UInt8>) -> ()]> of generic specialization <()> of Swift.StaticString.withUTF8Buffer<A>((Swift.UnsafeBufferPointer<Swift.UInt8>) -> A) -> A + 54
    frame #3: 0x00000001074d6c93 libswiftCore.dylib`partial apply forwarder for closure #2 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in Swift._fatalErrorMessage(Swift.StaticString, Swift.StaticString, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 99
    frame #4: 0x0000000107266856 libswiftCore.dylib`function signature specialization <Arg[1] = [Closure Propagated : reabstraction thunk helper from @callee_guaranteed (@unowned Swift.UnsafeBufferPointer<Swift.UInt8>) -> () to @callee_guaranteed (@unowned Swift.UnsafeBufferPointer<Swift.UInt8>) -> (@out ()), Argument Types : [@callee_guaranteed (@unowned Swift.UnsafeBufferPointer<Swift.UInt8>) -> ()]> of generic specialization <()> of Swift.StaticString.withUTF8Buffer<A>((Swift.UnsafeBufferPointer<Swift.UInt8>) -> A) -> A + 54
    frame #5: 0x00000001073eec40 libswiftCore.dylib`function signature specialization <Arg[2] = Dead, Arg[3] = Dead> of Swift._fatalErrorMessage(Swift.StaticString, Swift.StaticString, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 96
    frame #6: 0x0000000107266483 libswiftCore.dylib`Swift._fatalErrorMessage(Swift.StaticString, Swift.StaticString, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 19
  * frame #7: 0x000000010116bfaa Bond`BNDInvocation.readArgument<A>(index=3, self=0x00006040000383c0) at BNDInvocation.swift:60
    frame #8: 0x0000000101107212 Bond`closure #1 in ProtocolProxy.registerInvoker2<A, B, C>(invocation=0x00006040000383c0, block=0x00000001011137f0 Bond`partial apply forwarder for closure #1 (A, B) -> D in closure #1 (ReactiveKit.PublishSubject<C, ReactiveKit.NoError>) -> ReactiveKit.Disposable in Bond.ProtocolProxy.signal<A, B, C, D>(for: ObjectiveC.Selector, dispatch: (ReactiveKit.PublishSubject<C, ReactiveKit.NoError>, A, B) -> D) -> ReactiveKit.Signal<C, ReactiveKit.NoError> at ProtocolProxy.swift) at ProtocolProxy.swift:95
    frame #9: 0x00000001011129c5 Bond`partial apply for closure #1 in ProtocolProxy.registerInvoker2<A, B, C>(for:block:) at ProtocolProxy.swift:0
    frame #10: 0x0000000101105b5f Bond`thunk for @escaping @callee_guaranteed (@owned BNDInvocation) -> () at ProtocolProxy.swift:0
    frame #11: 0x0000000101112a21 Bond`thunk for @escaping @callee_guaranteed (@owned BNDInvocation) -> ()partial apply at ProtocolProxy.swift:0
    frame #12: 0x0000000101105b44 Bond`thunk for @escaping @callee_guaranteed (@in BNDInvocation) -> (@out ()) at ProtocolProxy.swift:0
    frame #13: 0x000000010110ce41 Bond`thunk for @escaping @callee_guaranteed (@in BNDInvocation) -> (@out ())partial apply at ProtocolProxy.swift:0
    frame #14: 0x0000000101105dc2 Bond`ProtocolProxy.handle(invocation=0x00006040000383c0, self=0x0000604000253590) at ProtocolProxy.swift:66
    frame #15: 0x0000000101105e3c Bond`@objc ProtocolProxy.handle(_:) at ProtocolProxy.swift:0
    frame #16: 0x00000001010d02cb Bond`-[BNDProtocolProxyBase forwardInvocation:](self=0x0000604000253590, _cmd="forwardInvocation:", invocation=0x00006040002741c0) at BNDProtocolProxyBase.m:86
    frame #17: 0x00007fff541425ac CoreFoundation`___forwarding___ + 748
    frame #18: 0x00007fff54142238 CoreFoundation`__forwarding_prep_0___ + 120
    frame #19: 0x00007fff517dd271 AppKit`-[NSOutlineView _expandItemEntry:expandChildren:startLevel:] + 417
    frame #20: 0x00007fff517d2cfb AppKit`-[NSOutlineView _uncachedNumberOfRows] + 313
    frame #21: 0x00007fff517d87ef AppKit`-[NSTableView _verifySelectionIsOK] + 517
    frame #22: 0x00007fff517d8558 AppKit`-[NSTableView _tileAndRedisplayAll] + 345
    frame #23: 0x00007fff518a4658 AppKit`-[NSOutlineView reloadData] + 586
    frame #24: 0x00007fff517d3a3e AppKit`-[NSOutlineView setDataSource:] + 261
    frame #25: 0x0000000100327464 Skala`key path setter for NSOutlineView.dataSource : <A>A.Base at NSOutlineView+Bond.swift:0
    frame #26: 0x000000010032749b Skala`___lldb_unnamed_symbol1540$$Skala + 27
    frame #27: 0x000000010733b585 libswiftCore.dylib`Swift.NonmutatingWritebackBuffer.__deallocating_deinit + 197
    frame #28: 0x00000001074f4120 libswiftCore.dylib`_swift_release_dealloc + 16
    frame #29: 0x000000010115c6b7 Bond`OptionalKeyPathProtocolProxyPropertyController.delegate.setter(newValue=0x0000604000253590, self=0x0000604000253860) at ProtocolProxyController.swift:0
    frame #30: 0x000000010110cd15 Bond`ProtocolProxy.registerDelegate(self=0x0000604000253590) at ProtocolProxy.swift:280
    frame #31: 0x0000000101106ff6 Bond`ProtocolProxy.registerInvoker2<A, B, C>(selector="outlineView:numberOfChildrenOfItem:", block=0x00000001011137f0 Bond`partial apply forwarder for closure #1 (A, B) -> D in closure #1 (ReactiveKit.PublishSubject<C, ReactiveKit.NoError>) -> ReactiveKit.Disposable in Bond.ProtocolProxy.signal<A, B, C, D>(for: ObjectiveC.Selector, dispatch: (ReactiveKit.PublishSubject<C, ReactiveKit.NoError>, A, B) -> D) -> ReactiveKit.Signal<C, ReactiveKit.NoError> at ProtocolProxy.swift, self=0x0000604000253590) at ProtocolProxy.swift:98
    frame #32: 0x000000010110a974 Bond`closure #1 in ProtocolProxy.signal<A, B, C, D>(subject=<unavailable>, self=0x0000604000253590, selector="outlineView:numberOfChildrenOfItem:", dispatch=0x0000000101111820 Bond`partial apply forwarder for closure #1 (ReactiveKit.PublishSubject<(), ReactiveKit.NoError>, B, C) -> D in Bond.ProtocolProxy.feed<A, B, C, D>(property: ReactiveKit.Property<A>, to: ObjectiveC.Selector, map: (A, B, C) -> D) -> ReactiveKit.Disposable at ProtocolProxy.swift) at ProtocolProxy.swift:200
    frame #33: 0x000000010110d075 Bond`partial apply for closure #1 in ProtocolProxy.signal<A, B, C, D>(for:dispatch:) at ProtocolProxy.swift:0
    frame #34: 0x0000000101109944 Bond`closure #1 in ProtocolProxy._signal<A>(observer=<unavailable>, registerInvoker=0x000000010110d020 Bond`partial apply forwarder for closure #1 (ReactiveKit.PublishSubject<C, ReactiveKit.NoError>) -> ReactiveKit.Disposable in Bond.ProtocolProxy.signal<A, B, C, D>(for: ObjectiveC.Selector, dispatch: (ReactiveKit.PublishSubject<C, ReactiveKit.NoError>, A, B) -> D) -> ReactiveKit.Signal<C, ReactiveKit.NoError> at ProtocolProxy.swift, self=0x0000604000253590, selector="outlineView:numberOfChildrenOfItem:") at ProtocolProxy.swift:148
    frame #35: 0x0000000101113949 Bond`partial apply for closure #1 in ProtocolProxy._signal<A>(for:registerInvoker:) at ProtocolProxy.swift:0
    frame #36: 0x00000001013e3c2b ReactiveKit`Signal.observe(observer=0x000000010138d570 ReactiveKit`partial apply forwarder at Subjects.swift, self=(producer = 0x0000000101113930 Bond`partial apply forwarder for closure #1 (ReactiveKit.AtomicObserver<A, ReactiveKit.NoError>) -> ReactiveKit.Disposable in Bond.ProtocolProxy.(_signal in _23028F409340362020B24E8EAAA251BA)<A>(for: ObjectiveC.Selector, registerInvoker: (ReactiveKit.PublishSubject<A, ReactiveKit.NoError>) -> ReactiveKit.Disposable) -> ReactiveKit.Signal<A, ReactiveKit.NoError> at ProtocolProxy.swift)) at Signal.swift:41
    frame #37: 0x00000001013e3e97 ReactiveKit`protocol witness for SignalProtocol.observe(with:) in conformance Signal<A, B> at Signal.swift:0
    frame #38: 0x00000001013a2dba ReactiveKit`SignalProtocol.observe<A>(observer=<unavailable>, self=) at SignalProtocol.swift:50
    frame #39: 0x00000001013a0374 ReactiveKit`ConnectableSignal.connect(self=0x0000604000058c60) at Connectable.swift:50
    frame #40: 0x00000001013a0630 ReactiveKit`protocol witness for ConnectableSignalProtocol.connect() in conformance ConnectableSignal<A> at Connectable.swift:0
    frame #41: 0x00000001013a0be2 ReactiveKit`closure #1 in ConnectableSignalProtocol.refCount(observer=0x00006040000e4080, lock=0x00006040000d5070, count=0x0000604000036db0, self=0x0000604000058c60, connectionDisposable=nil) at Connectable.swift:76
    frame #42: 0x00000001013a0e12 ReactiveKit`partial apply for closure #1 in ConnectableSignalProtocol.refCount() at Connectable.swift:0
    frame #43: 0x00000001013e3c2b ReactiveKit`Signal.observe(observer=0x00000001011118c0 Bond`reabstraction thunk helper from @escaping @callee_guaranteed (@unowned ReactiveKit.Event<(), ReactiveKit.NoError>) -> () to @escaping @callee_guaranteed (@in ReactiveKit.Event<(), ReactiveKit.NoError>) -> ()partial apply forwarder with unmangled suffix ".76" at ProtocolProxy.swift, self=(producer = 0x00000001013a0dd0 ReactiveKit`partial apply forwarder for closure #1 (ReactiveKit.AtomicObserver<A.Element, A.Error>) -> ReactiveKit.Disposable in (extension in ReactiveKit):ReactiveKit.ConnectableSignalProtocol.refCount() -> ReactiveKit.Signal<A.Element, A.Error> at Connectable.swift)) at Signal.swift:41
    frame #44: 0x000000010110dd6a Bond`ProtocolProxy.feed<A, B, C, D>(property=<unavailable>, selector="outlineView:numberOfChildrenOfItem:", map=0x00000001010f5d70 Bond`partial apply forwarder for reabstraction thunk helper <A where A: ReactiveKit.SignalProtocol, A.Element: Bond.ObservableCollectionEventProtocol, A.Error == ReactiveKit.NoError, A.Element.UnderlyingCollection: Bond.TreeNodeProtocol> from @escaping @callee_guaranteed (@in Swift.Optional<A.Element.UnderlyingCollection>, @owned __ObjC.NSOutlineView, @in Swift.Optional<A.Element.UnderlyingCollection>) -> (@unowned Swift.Int) to @escaping @callee_guaranteed (@in Swift.Optional<A.Element.UnderlyingCollection>, @in __ObjC.NSOutlineView, @in Swift.Optional<A.Element.UnderlyingCollection>) -> (@out Swift.Int) at NSOutlineView+ObservableTree.swift, self=0x0000604000253590) at ProtocolProxy.swift:305
    frame #45: 0x00000001010f47b6 Bond`SignalProtocol<>.bind(outlineView=0x000000010a062ab0, binder=<unavailable>, self=0x00007ffeefbfd760) at NSOutlineView+ObservableTree.swift:126
    frame #46: 0x00000001000aea23 Skala`PageLayerOrganizerViewController.viewDidLoad(self=0x00006000001230c0) at PageLayerOrganizerViewController.swift:81
    frame #47: 0x00000001000af3b4 Skala`@objc PageLayerOrganizerViewController.viewDidLoad() at PageLayerOrganizerViewController.swift:0
    frame #48: 0x00007fff5172d20a AppKit`-[NSViewController _sendViewDidLoad] + 97
    frame #49: 0x00007fff51cd1afd AppKit`_noteLoadCompletionForObject + 641
    frame #50: 0x00007fff516b197d AppKit`-[NSIBObjectData nibInstantiateWithOwner:options:topLevelObjects:] + 2042
    frame #51: 0x00007fff517acd1e AppKit`-[NSNib _instantiateNibWithExternalNameTable:options:] + 679
    frame #52: 0x00007fff517ac97a AppKit`-[NSNib _instantiateWithOwner:options:topLevelObjects:] + 136
    frame #53: 0x00007fff517abc50 AppKit`-[NSViewController loadView] + 343
    frame #54: 0x00007fff51723a9e AppKit`-[NSViewController _loadViewIfRequired] + 75
    frame #55: 0x00007fff51723a09 AppKit`-[NSViewController view] + 30
    frame #56: 0x000000010003b2a8 Skala`SlidingSeguePresentationAnimator.swapViewController(outgoingViewController=0x00006000001215e0, incomingViewController=0x00006000001230c0, presenting=true, self=0x0000600000239520) at SlidingSegue.swift:94
    frame #57: 0x000000010003b0c2 Skala`SlidingSeguePresentationAnimator.animatePresentation(viewController=0x00006000001230c0, fromViewController=0x00006000001215e0, self=0x0000600000239520) at SlidingSegue.swift:86
    frame #58: 0x000000010003b131 Skala`@objc SlidingSeguePresentationAnimator.animatePresentation(of:from:) at SlidingSegue.swift:0
    frame #59: 0x000000010003a2d1 Skala`SlidingSegue.perform(self=0x000060000064c840) at SlidingSegue.swift:16
    frame #60: 0x000000010003a404 Skala`@objc SlidingSegue.perform() at SlidingSegue.swift:0
    frame #61: 0x00007fff51b76256 AppKit`-[NSStoryboardShowSegueTemplate _perform:] + 275
    frame #62: 0x00007fff51bd615f AppKit`-[NSViewController performSegueWithIdentifier:sender:] + 294
    frame #63: 0x000000010014bb46 Skala`PageOrganizerViewController.showPage(sender=0x000000010a248af0, self=0x00006000001215e0) at PageOrganizerViewController+Actions.swift:66
    frame #64: 0x000000010014bbac Skala`@objc PageOrganizerViewController.showPage(_:) at PageOrganizerViewController+Actions.swift:0
    frame #65: 0x00007fff51e79a43 AppKit`-[NSApplication(NSResponder) sendAction:to:from:] + 312
    frame #66: 0x00007fff5191f53f AppKit`-[NSControl sendAction:to:] + 86
    frame #67: 0x00007fff51993069 AppKit`-[NSTableView _sendAction:to:row:column:] + 121
    frame #68: 0x00007fff519914a8 AppKit`-[NSTableView mouseDown:] + 7186
    frame #69: 0x00007fff52018d6d AppKit`-[NSWindow(NSEventRouting) _handleMouseDownEvent:isDelayedEvent:] + 5891
    frame #70: 0x00007fff520159c4 AppKit`-[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:] + 2359
    frame #71: 0x00007fff52014c70 AppKit`-[NSWindow(NSEventRouting) sendEvent:] + 497
    frame #72: 0x00007fff51e76236 AppKit`-[NSApplication(NSEvent) sendEvent:] + 2462
    frame #73: 0x00007fff516d68b5 AppKit`-[NSApplication run] + 812
    frame #74: 0x00007fff516a5a72 AppKit`NSApplicationMain + 804
    frame #75: 0x000000010026d34d Skala`main at ApplicationDelegate.swift:10
    frame #76: 0x00007fff7c461015 libdyld.dylib`start + 1
```

Reply:
So a bit of further info on this crash (I haven't figured it out yet):

It's crashing when trying to provide:

```swift
func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int
```

when trying to `readArgument()` at index 3 (I assume that's the result?). It's crashing at `BNDInvocation.swift:L60` with the following error:

```
Thread 1: Fatal error: Unexpectedly found nil while unwrapping an Optional value
```

Do you have any ideas of how I can track this down? My local implementation of that method in the OutlineViewBinder is:

```swift
disposable += outlineView.reactive.dataSource.feed(
    property: dataSource,
    to: #selector(NSOutlineViewDataSource.outlineView(_:numberOfChildrenOfItem:)),
    map: { (dataSource: Element.UnderlyingCollection?, _: NSOutlineView, item: Element.UnderlyingCollection?) -> Int in
        guard let dataSource = dataSource else { return 0 }
        return item?.children.count ?? dataSource.count // TODO check if this is correct
    }
)
```

Reply:
Thanks for the sample Tony! This was an issues in BNDInvocation and should be fixed now. There are few issues in OutlineViewBinder though. I'm going to make few comments.

Reply:
> Maybe we should refactor this so that the root node is not displayed in the outline view, rather the first level is root node's children.

Yes, I agree with this. We should remove the requirement for a value on the root node. I'm not sure how this would look.

Reply:
> From what I can see the binder calls appropriate methods on the outline view but nothing happens.

I've been looking into this today. I'm honestly not sure why these changes aren't propagating to the `NSOutlineView` control. The events appear to be coming through as expected, but calling the insert/etc methods on the outline view have no effect. 

> Maybe we should refactor this so that the root node is not displayed in the outline view.

I've pushed up a change that _should_ do just that.

Reply:
I'll be working on this tomorrow. I'm thinking of introducing another type that will act as a root node for such trees. Something like `TreeArray` that would contain an array of `TreeNode`s. `TreeArray` is what we would bind to NSOutlineView then. It'd basically be a `TreeNode` without a value.

Reply:
Sounds perfect, and more closely matches the situation I'm working with - thanks @srdanrasic! 

I've had no luck so far working out why the outline view is not updating. I also have a cold. Extrapolate from that what you will 😜